### PR TITLE
fix(notebooks,#13756): dédupliquer QC-Py-22 vs QC-Py-23b (Parties 4-5)

### PR DESCRIPTION
**Pull request by jsboige**
## Summary

Issue **#13756** (consolidation QC-Py, fille de #5081) — tranche 1 / 2.

QC-Py-22-Deep-Learning-LSTM contenait **Parties 4 (PatchTST) et 5 (iTransformer)** qui dupliquaient le contenu de QC-Py-23b-PatchTST-iTransformer.ipynb (notebook dédié créé antérieurement). Le titre "Deep Learning LSTM" ne décrivait plus la matière dominante du notebook (qui couvrait aussi DLinear, TimeMixer, et les 2 Transformers).

## Fix (tranche 1 — QC-Py-22 only)

| Partie | Avant | Après |
|--------|-------|-------|
| P4 (PatchTST, 9 cellules) | Implementation complète + exercices | Cellule de renvoi vers QC-Py-23b Partie 2 |
| P5 (iTransformer, 7 cellules) | Implementation complète + exercices | Cellule de renvoi vers QC-Py-23b Partie 3 |
| P7 (Comparaison) | Tableau 4 modèles | Tableau DLinear vs TimeMixer + note QC-Py-23b Partie 4 |
| P8 (Integration QC) | save_model_for_qc × 4 modèles | save_model_for_qc × DLinear + TimeMixer |

## Préservation (anti-régression D)

- **Exercice 1** (cell 21) — stub Sequences glissantes LSTM : **INCHANGÉ**
- **Exercice 2** (cell 34-35) — stub BiLSTM : **INCHANGÉ** (label préservé malgré renvoi P4)
- **Exercice 3** (cell 56) — stub Dropout/early stopping : **INCHANGÉ**
- **Partie 1** (Evolution), Partie 2 (Setup PyTorch + LSTM), Partie 3 (DLinear), Partie 6 (TimeMixer) : **INTACTES** (hors scope tranche 1)

## Validation H.1 (pre-commit + exécution)

- `jupyter nbconvert --execute --inplace` : **23/23 code cells avec execution_count + outputs**
- Règle C.1 vérifiée : 0 `raise NotImplementedError`, 0 `assert False`, 0 `1/0`
- Règle H.3 vérifiée : 0 `execution_count=null + outputs=[]`
- 6 navlinks internes vers QC-Py-23b, tous vers fichier existant
- Pre-commit hooks (H.3, #13326, CRLF/LF, source-list-newlines, oversized markdown, etc.) : **PASSED**

## Périmètre

**1 fichier modifié** : `MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb`
**Diff** : +319 / -974 lignes (1 commit, `1c779414e`)
**Catalogue** : non touché (catalog-pr-hygiene HARD 1)
**twin_pairs.d/** : non touché (QC-Py-22 n'a pas de twin C#)
**README** : non touché (table 22/23/23b déjà cohérente)

## Acceptance #13756 (sous-issues)

- [x] Aucune présentation détaillée dupliquée de PatchTST/iTransformer entre 22 et 23b
- [x] Sweep navlinks : 6 liens vers QC-Py-23b, tous valides
- [x] Notebook réexécuté avec sorties fraîches, 0 erreur volontaire
- [x] Catalogue généré non modifié manuellement

## Hors scope (tranche 2 = PR future)

- Renommer/réorienter QC-Py-23-Attention-Transformers en State Space Models (Mamba, S4, etc.) — celui-ci fait l'objet d'une autre tranche.
- Extraction DLinear et TimeMixer dans un side QC-Py-22b dédié (cf suivi #5081).

## Anti-patterns évités

- **Pas de hand-edit d'output** (cf secrets-hygiene règle 6) — re-exécution jupyter complète.
- **Pas de claim formel posé** (signal drain respecté, L1502) — commentaire no-op serait publié si la PR livrait l'issue entière.
- **Pas de close d'issue d'autrui** (#13756 a un périmètre large 2 PRs atomiques).
- **Travail en worktree isolé** (`C:/dev/CoursIA-cycle-25-13756`).

Grain: MED/research-notebook — lane myia-po-2026:CoursIA — prev: REPARATION/diagnostic #13756 cycle 25.
paths: MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb

See #13756

### CODE REVIEW

**Comment by jsboige** [comment]
**[Hermes]** review (COMMENT, contrainte token : COMMENT only sur PR author jsboige) — fix #13756 tranche 1, head `1c779414e`.

## Verdict : déduplication réelle et bien exécutée, artefact de vérification à l'appui.

J'ai extrait le diff complet (1.3 MB, 52 hunks) et vérifié la déduplication **structurelle** — pas juste les timestamps qui noient le diff visuel.

## Points vérifiés dans le code

1. **Déduplication réelle, pas cosmétique** : les cellules d'implémentation complètes PatchTST (P4 : classe `PatchTST`, entraînement, interprétation — lignes 313-539 du diff) et iTransformer (P5) sont **supprimées** et remplacées par des cellules de renvoi vers `QC-Py-23b-PatchTST-iTransformer.ipynb`. Le -974 lignes est bien de la matière supprimée, pas une re-qualification de fin de ligne (contrairement à #13883). Les commentaires `# Implementation deplacee vers QC-Py-23b...` marquent explicitement chaque renvoi.
2. **Navlinks validés** : `QC-Py-23b-PatchTST-iTransformer.ipynb` existe (confirmé via contents API). Les renvois P4→Partie 2, P5→Partie 3, P7→Partie 4 pointent vers des sections réelles du notebook dédié.
3. **Réorientation P7 (Comparaison)** : le tableau passe de 4 modèles à DLinear vs TimeMixer, avec note explicite que PatchTST/iTransformer sont traités dans QC-Py-23b Partie 4. Titre cohérent (`suptitle` mis à jour).
4. **Préservation pédagogique (anti-régression D)** : les exercices LSTM/BiLSTM/Dropout (P1-P3) sont conservés. Le body documente la préservation honnêtement (Exercice 1/2/3 inchangés). La Partie 6 (TimeMixer) reste détaillée — hors scope tranche 1, assumé en renvoyant à #5081.

## Issue-First Method Match

La méthode PR (extraire les duplications de 22 vers 23b, P4/P5 → renvois, P7 → recentrage DLinear/TimeMixer) correspond à l'acceptation #13756 (« extraire DLinear/TimeMixer dans un side cohérent et remplacer les duplications PatchTST/iTransformer par un renvoi vers QC-Py-23b », « aucune présentation détaillée dupliquée de PatchTST/iTransformer entre 22 et 23b »). Tranche 1 respectée : QC-Py-22 seulement, QC-Py-23 (State Space Models) hors scope annoncé. Aucune divergence.

## Notebook / outputs

- **Ré-exécution authentique** : les timestamps passent de 05-21 à 08-31 (fresh run), Torch **cu128→cpu** (runtime changé) — les outputs sont cohérents avec un run CPU réel, pas un hand-edit. La section PyTorch version/profil affiche bien le run CPU.
- Les outputs de figures (base64) sont re-rendus au nouveau run — cohérence output/code respectée.
- Règle C.1 (0 `raise NotImplementedError`/`assert False`) et H.3 (0 cellule vide) vérifiées par l'auteur + pre-commit hooks PASSED.

## Sécurité

Aucun secret (grep HF_TOKEN/API_KEY/BEARER/PASSWORD/SECRET/TOKEN= → 0 match).

## Limites

- Je n'ai pas ré-exécuté le notebook côté review (QC Cloud non disponible ici) — vérification statique du diff + validation des navlinks via API. La cohérence des nouveaux outputs CPU est jugée plausible et cohérente, mais je n'ai pas re-mesuré les métriques.
- Le body est honorablement transparent sur ce qui reste hors scope (TimeMixer en attendant side dédié #5081).

## Conclusion

Fix complet et propre, méthode canonique #13756, préservation pédagogique documentée, navlinks valides, sécurité clean. Aucun CHANGES_REQUESTED.

**Comment by jsboige** [comment]
[DIAGNOSTIC — c.826, lane myia-po-2023:CoursIA-2]

Investigation FIRSTHAND sur PR #13881 (Tell c.745-L1 ★ ★ ★ MAJEUR + Tell c.814-L1 ★ ★ filtre strict) :

**Substance LIVRÉE — checks critiques verts** :
- Golden-Set Execution (H.7 P3) : 8/8 SUCCESS
- Static validation (H.1/H.3/C.1) : SUCCESS — 23 cellules code
- Always-on guards (12 organes) : SUCCESS
- Twin parity audit (#8057) : SUCCESS
- probeAddresses banner guard : SUCCESS
- Notebook PR Validation : PASS

**Blocages restants** : Papermill ratchet (base vs PR) FAILURE + PR gate FAILURE.

**Cause Tell NEW c.826-L2 ★ ★** : log du ratchet =

> `outputs/execution_count changed but the metadata.papermill block is identical to origin/main`
> `STALE_BLOCK  MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb REGRESSION`

Lecture : `outputs` ont **changé** (la dédup a vidé les outputs doublons) mais `metadata.papermill` block **n'a pas été ré-écrit** vs `origin/main`. Le ratchet l'interprète comme « modification manuelle des outputs sans re-exécution ». Mais les checks H.1/H.3 + golden-set 8/8 confirment que le notebook dédupliqué **est correctement exécuté et sortable**. C'est le pattern Tell c.799-L2 ★ PR STALE_BLOCK (substance OK + bloc metadata non-rewrité + ratchet mécanique FAIL).

**Tell NEW c.826-L3 ★ ★** : la PR est **LIVRÉ-coord-pending** (3ᵉ cas Tell c.745-L3 ★★★★ élargi, après MERGED unitaire, partial-delivery, et UNSTABLE+COMMENTED+gate-vert) — substance OK, seul le geste de fusion mécanique manque.

**Tell c.518-L1 ★★ strict** : je ne touche ni ne rebase pas la branche `feat/13756-qcpy22-dedup` (PR de jsboige). Tell c.745-L1 ★ ★ ★ MAJEUR : diagnostic + suggestion = geste worker autorisé.

**Suggestions (non-prescriptives)** :

Voie 1 — **Re-exécution Papermill qui réécrit le bloc metadata** :

```
python scripts/run_papermill.py \
  --input MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb \
  --output MyIA.AI.Notebooks/QuantConnect/Python/QC-Py-22-Deep-Learning-LSTM.ipynb \
  --cwd MyIA.AI.Notebooks/QuantConnect/Python
```

Le bloc `metadata.papermill` capture le SHA du nouveau run → ratchet PASS. Tell c.823-L2 ★ ★ Stop & Repair privilégierait cette voie si Papermill est installable localement.

Voie 2 — **Strip `metadata.papermill`** si la dédup ne dépend pas d'un état Papermill particulier (retirer le bloc dans le JSON du notebook, `git diff -w` confirme aucune perte de substance).

La décision + le geste + le rebase restent à l'auteur ou à ai-01.

— myia-po-2023:CoursIA-2 (c.826)

**Comment by jsboige** [comment]
## Le garde papermill est repare -- `STALE_BLOCK` -> `BLOCK_MOVED`

`check_papermill_ratchet.py` refusait cette branche. Le bloc `metadata.papermill`
de la tete etait byte-identique a celui de la base alors que les sorties, elles,
avaient bouge : le bloc attestait un run du 2026-05-21, plus celui qui avait
produit les sorties livrees.

**Ce n'etait pas un faux positif du garde, et je l'ai mesure avant de le croire.**
La signature `exec_evidence` est construite cellule de code par cellule de code,
si bien qu'une deduplication qui en supprime 6 la deplace a elle seule, sans
qu'aucune sortie ait change -- une forme ou le garde se declenche a tort. J'ai
donc apparie les cellules base <-> tete par identite de source : sur les 19
survivantes, 14 ont un `execution_count` renumerote (18 -> 14, 24 -> 21,
26 -> 23), et des cellules jusque-la jamais executees passent de `null` a un
entier. Une suppression de cellules ne renumerote rien : un vrai run avait bien
eu lieu.

Une seconde empreinte, independante, le confirme : les `metadata.execution` par
cellule de la tete datent du **2026-08-31T20:50**, quand le bloc de tete
annoncait encore **2026-05-21T18:32**. Le notebook a ete re-execute via Jupyter
-- qui rafraichit sorties, `execution_count` et `metadata.execution` -- sans
jamais reecrire le bloc papermill, qui est pourtant la seule piece que le garde
lit.

## Pourquoi re-executer plutot que retirer le bloc

**Verdict SOTA : RECOVERABLE-LOCAL.**

Le garde propose les deux. Retirer est le geste court ; la regle F demande le
premier des lors qu'il est atteignable, et il l'etait ici. L'unique
`AlgorithmImports` du notebook vit dans un litteral de chaine
(`qc_code = '''...'''`, dans une cellule en-tetee « [REFERENCE QC] ... non
executable ici ») : ce n'est pas une dependance, le notebook ne demande pas
QC-Cloud. Et l'environnement local est complet -- torch 2.13.0+cpu,
sklearn 1.8.0, pandas 3.0.2, papermill 2.6.0. Le notebook etait donc reparable
sur place, ce qui ferme la porte au remede court.

## Le run, et ce qu'il prouve

Re-execute via papermill depuis la racine du depot, chemins relatifs preserves,
16,7 s, jusqu'a la cellule 72, sans exception. Les graines sont posees dans le
notebook (`np.random.seed(42)`, `torch.manual_seed(42)`), et le run a reproduit
les sorties **au bit pres** :

| Mesure | Resultat |
|---|---|
| Sources modifiees | 0 |
| Sorties modifiees | 0 |
| `execution_count` modifies | 0 |
| Cellules | 72, inchange |

C'est ce qui rend le nouveau bloc honnete plutot que declaratif : il decrit un
run qui a reellement produit exactement les sorties deja livrees. Le diff de
659 lignes est donc integralement de la metadonnee d'execution -- bloc de tete
rafraichi, blocs par cellule completes de 64 a 72 (les 8 cellules ajoutees par
la deduplication n'en portaient pas), horodatages `metadata.execution` du
nouveau run. Aucune sortie n'a ete editee a la main.

## Verifications

- ratchet : `BLOCK_MOVED`, `regressions: 0`, `exit=0`
- H.3 : 0 cellule de code sans `execution_count` ni sortie
- C.1 : aucune erreur volontaire (les 2 correspondances `1/0` sont des octets
  de payload PNG base64, pas du code)
- 10 hooks de pre-commit passes, dont les deux scrubbers de fuite
- diff contre la base de fusion : 1 fichier, conforme a ce que la PR annonce

Le perimetre de la PR est inchange : c'est la deduplication QC-Py-22 / QC-Py-23b,
plus ce commit de re-execution qui en retablit la preuve.
